### PR TITLE
AspNetCoreVersion 2.0, re-done from the last master.

### PR DESCRIPTION
**Pull request by jtkech**
So, i've used a `Visual Studio Preview` version and `AspNetCoreVersion  = 2.0.0-*`.

Orchard.ResponseCache
---------------------------
- `IResponseCachingPolicyProvider` interface has more methods to override.

ThemingViewsFeatureProvider
---------------------------
- Class `ViewsFeature` has changed, new `CompiledViewDescriptor` class.

SharedCompilerCacheProvider
---------------------------
- No more `ICompilerCacheProvider`, new `IViewCompilerProvider`, new `RazorViewCompiler` class.
- Becomes a `SharedViewCompilerProvider` where we cache only `ViewDescriptors`.

Orchard.Setup
-------------
- Remove `@using Microsoft.AspNetCore.Builder`.
-  Add `@using Microsoft.AspNetCore.Identity`.
- Use `Layout = "../_layout.cshtml"`

Orchard.Admin
-------------
- Add `[Authorize]` attribute to the `AdminController` (was not necessary before).

Orchard.Security
----------------
- New `AuthorizationResult` class.

Orchard.Users
-------------
- `UseIdentity()` is obsolete, `IdentityOptions` has changed.
- `UseCookieAuthentication()` is deprecated.
- No more `IdentityMarkerService`, no more `IdentityCookieOptions`.
- Use `HttpContext.SignOutAsync()` in place of `HttpContext.Authentication.SignOutAsync()`.

Orchard.OpenId
--------------
- No more `UseOpenIddict()`, `UseJwtBearerAuthentication()` is deprecated.
- Just commented out lines in the `Startup` to be able to build and run.
- `ActiveAuthenticationSchemes` => `AuthenticationSchemes`.
- To be able to build: `HttpContext.Authentication.GetAuthenticateInfoAsync()` => `HttpContext.AuthenticateAsync()`.

ExtensionManagerTests
---------------------
- No more `PlatformServices`.
- `PlatformServices.Default.Application.ApplicationBasePath` => `AppContext.BaseDirectory`.

MenuPart.Edit.cshtml
--------------------
- `@model` and `@inject` directives at the start of a line and also before any html markup.

ContentFileInfo
---------------
- `public string PhysicalPath => null`;

TheTheme\Views\Layout.cshtml
TheAdmin\Views\Layout.cshtml
----------------------------
- `@inject` directive was at the start of the line, but apparently with a bad `@` character. I needed to rewrite this character and save, not needed for the `TheBlogTheme\Views\Layout.cshtml`.

VStudio
-------
- Before, when starting the app the target `AfterBuild` was always called, and then assets files was always copied. Now this is not the case if the application doesn't need to be built.

- Not already checked but i think it's ok when only updating an asset file. This because now, during development, we point asset files directly in the referenced projects.

- If you just add / remove an asset file, i think you need to force a build. E.g by doing a `ReBuild` or, to be quicker and just to build the app project, by only updating one of its file (e.g `Startup.cs`).

- And for publishing i think that as before there is in a 1st step a full building.

1st tests with CMS recipe
----------
- Setup + login + admin + new page + new menu item.

1st tests with Blog recipe
----------
- Setup + login + admin + new page + new menu item.

Unit tests
----
- They pass on my machine.

### CODE REVIEW

**Comment by jtkech** [comment]
@alexbocharov Thanks.

**Comment by alexbocharov** [comment]
:+1:

**Comment by alexbocharov** [comment]
@sebastienros @jtkech @Jetski5822  ImageSharp.Web blocks transition to .NET Core 2.0

**Comment by sebastienros** [comment]
I can rebuild it I think.

**Comment by jtkech** [comment]
Can't work on it today but looks good 👍

File: src/OrchardCore.Modules/Orchard.OpenId/Startup.cs [hunk 1]
```diff
@@ -42,6 +42,7 @@ public class Startup : StartupBase
 
         public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
         {
+            /*
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/Orchard.OpenId/Startup.cs]: Hum, this makes the OpenID module completely unusable. I'll try to take a look before you merge it.

Comment (jtkech) [inline src/OrchardCore.Modules/Orchard.OpenId/Startup.cs]: Yes, as mentioned in my above list about `Orchard.OpenId`.

    Just commented out lines in the Startup to be able to build and run

So, for this part i didn't do anything 😄 

In my last comment, when i said looks good it was related to the last work of @alexbocharov.

**Comment by PinpointTownes** [comment]
@alexbocharov @jersiovic @jtkech @sebastienros I spent some time today porting the OpenID module to ASP.NET Core 2.0 and it now works like a charm (I only tested the password flow + the JWT/validation handlers integration, tho'): https://github.com/OrchardCMS/Orchard2/pull/887/commits/22fb7f10de95974e4a30ba997b760f8f134d973f

Note: there's a lot of infrastructure code, but it's unfortunately unavoidable due to the massive changes in ASP.NET Core 2.0's authentication stack, that now require configuring everything in `ConfigureServices()`.

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdConfiguration.cs [hunk 2]
```diff
@@ -1,35 +1,102 @@
...
+            if (settings.AccessTokenFormat == OpenIdSettings.TokenFormat.Encrypted)
+            {
+                options.AddScheme(OAuthValidationDefaults.AuthenticationScheme, builder =>
+                {
+                    builder.HandlerType = typeof(OAuthValidationHandler);
+                });
+            }
+            else if (settings.AccessTokenFormat == OpenIdSettings.TokenFormat.JWT)
+            {
+                // Note: unlike most authentication handlers in ASP.NET Core 2.0,
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdConfiguration.cs]: @HaoK FYI.

Comment (alexbocharov) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdConfiguration.cs]: @PinpointTownes Thanks.

**Comment by alexbocharov** [comment]
@PinpointTownes Thanks.

**Comment by jtkech** [comment]
@PinpointTownes 👍

**Comment by jtkech** [comment]
@alexbocharov 

- About `ImageSharp.Web`, yes just saw this. Before they target 2.0 or seb rebuild it, the temporary workaround i found is to reference `"System.Numerics.Vectors" Version="4.4.0"` directly in `Orchard.Media`.

- Then, when using media i got some errors (maybe other issues). But at least we can build and run.

**Comment by alexbocharov** [comment]
@jtkech 
Will watch

**Comment by sebastienros** [comment]
I updated the packages for imagesharp (also fixing a bug) and yessql for netcore20.

**Comment by alexbocharov** [comment]
Ok, I will update

File: src/OrchardCore.Build/AspNetCoreVersion.props [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <PropertyGroup>
```
Comment (PinpointTownes) [inline src/OrchardCore.Build/AspNetCoreVersion.props]: Curious: have you consider renaming this file and adding all the external dependencies versions? (e.g Fluid, YesSql, OpenIddict, etc.)

That would be a great way to have a centralized place where versions are defined.

**Pull request merged**